### PR TITLE
Fb import

### PR DESCRIPTION
**Pull request by philiphand**

### CODE REVIEW

File: src/custom_modules/DAO_modules/donations.js [hunk 1]
```diff
@@ -865,7 +898,9 @@ module.exports = {
     getYearlyAggregateByDonorId,
     getByDonorId,
     getLatestByKID,
+    GetByExternalPaymentID,
```
Comment (fellmirr) [inline src/custom_modules/DAO_modules/donations.js:901]: These have wrong casing (should start with a lowercase)

File: src/custom_modules/DAO_modules/donors.ts [hunk 2]
```diff
@@ -100,6 +100,41 @@ async function getByKID(KID): Promise<Donor | null> {
...
+      // Query above does not find donors that have not donated before
+      if(result.length == 0) {
+          [result] = await con.execute(`
+              SELECT ID FROM Donors
+              where full_name = ?
+          `, [name])
+      }
+
+      con.release()
+      if (result.length > 0) return (result[0].ID)
```
Comment (fellmirr) [inline src/custom_modules/DAO_modules/donors.ts:129]: What if there are multiple donors with the same full name?

File: src/custom_modules/DAO_modules/facebook.js [hunk 3]
```diff
@@ -1,5 +1,23 @@
 var con
 
+//region Get
+
+async function getAllFacebookDonations(donorID, paymentID) {
+    try {
+        let [results] = await con.query(`
+            SELECT PaymentExternal_ID, sum_confirmed, timestamp_confirmed
+            FROM Donations
+            WHERE Payment_ID = 9`
```
Comment (fellmirr) [inline src/custom_modules/DAO_modules/facebook.js:10]: Pull the number from the paymentMethod enum in `/enums/paymentMethods.js`

File: src/routes/distributions.js [hunk 4]
```diff
@@ -94,7 +94,9 @@ router.get("/:KID",
   }
 })
 
-router.get("/:KID/unauthorized", async (req, res, next) => {
+router.get("/:KID/unauthorized",
+  authMiddleware(authRoles.read_all_donations),
```
Comment (fellmirr) [inline src/routes/distributions.js:98]: Authrole here should be .admin

File: src/routes/distributions.js [hunk 5]
```diff
@@ -104,6 +106,35 @@ router.get("/:KID/unauthorized", async (req, res, next) => {
   }
 })
 
+router.post("/KID/distribution",
+  authMiddleware(authRoles.read_all_donations),
```
Comment (fellmirr) [inline src/routes/distributions.js:110]: Same here, should have auth role .admin

File: src/routes/facebook.js [hunk 6]
```diff
@@ -3,13 +3,31 @@ const express = require('express')
...
+const roles = require('../enums/authorizationRoles')
 
 function throwError(message) {
     let error = new Error(message)
     error.status = 400
     throw error
 }
 
+router.get("/payments/all",
+    auth(roles.read_all_donations),
```
Comment (fellmirr) [inline src/routes/facebook.js:16]: Should use `authMiddleware.isAdmin`

**Review by fellmirr — CHANGES_REQUESTED** [review]

File: src/custom_modules/DAO_modules/donors.ts (hunk 2 above)
Comment (philiphand) [reply src/custom_modules/DAO_modules/donors.ts:129]: It intentionally selects the donor who most recently donated (see the SQL query), so only the first donor should be returned

**Pull request merged**